virttest/utils_qemu.py

```diff
@@ -148,3 +154,128 @@ def get_maxcpus_hard_limit(bin_path, machine_type):
+            cmd_dict['port'] = '-p %s' % params['nbd_port']
+
+        # tls creds is supported for tcp only
+        if params.get('nbd_server_tls_creds'):
+            tls_str = '--object tls-creds-x509,id={aid},endpoint=server,dir={tls_creds} --tls-creds {aid}'
+            cmd_dict['tls_creds'] = tls_str.format(
+                aid='%s_server_tls_creds' % local_image,
+                tls_creds=params['nbd_server_tls_creds']
+            )
+
+    cmdline = qemu_nbd + ' ' + string.Formatter().format(export_cmd,
+                                                         **cmd_dict)
+    return process.run(cmdline, ignore_status=True, shell=True,
+                       ignore_bg_processes=True)
+
+
+def list_image_with_qemu_nbd(qemu_nbd, nbd_image, params):
+    """
+    List all details about the exports by qemu-nbd
+    :param qemu_nbd: path to the qemu-nbd binary
+    :param nbd_image: nbd image tag defined in parameter images
+    :param params: nbd image specified params
+    :return: CmdResult object of qemu-nbd output
+    """
+    cmd_dict = {
+        "tls_creds": "",
+        "port": "",
+        "server": ""
+    }
+    list_cmd = '-L {tls_creds} {port} {server}'
+
+    if params.get('nbd_unix_socket'):
+        cmd_dict['server'] = '-k %s' % params['nbd_unix_socket']
+    else:
+        # 0.0.0.0 is used by default if interface is not set
+        if params.get('nbd_server'):
+            cmd_dict['server'] = '-b %s' % params['nbd_server']
+
+        # 10809 is used by default if port is not set
+        if params.get('nbd_port'):
+            cmd_dict['port'] = '-p %s' % params['nbd_port']
+
+        # tls creds is supported for tcp only
+        tls = storage.StorageAuth.auth_info_define_by_params(nbd_image, params)
+        if tls:
+            tls_str = '--object tls-creds-x509,id={t.aid},endpoint=client,dir={t.tls_creds} --tls-creds {t.aid}'
+            cmd_dict['tls_creds'] = tls_str.format(t=tls)
+
+    cmdline = qemu_nbd + ' ' + string.Formatter().format(list_cmd,
+                                                         **cmd_dict)
+    return process.run(cmdline, ignore_status=True, shell=True,
+                       ignore_bg_processes=True)
+
+
+def stop_qemu_nbd(image_params):
+    """
+    Stop a qemu-nbd process which exports an image
+    :param image_params: image specified params
+    :return: CmdResult object of kill output
+    """
+    # Exactly match the nbd port or unix domain socket
+    match = image_params['nbd_unix_socket'] if image_params.get(
+        'nbd_unix_socket') else image_params.get('nbd_port')
+
+    # Neither port nor unixsock is set, the default port must be used
+    none_match = '\-p|\--port|\-k|\--socket' if not match else ''
+
+    filter_cmd = 'grep "%s"' % match if match else 'egrep -v "%s"' % none_match
+    kill_cmd = "kill -9 $(ps aux | grep 'qemu-nbd' | %s | awk '{print $2}')" % filter_cmd
+    return process.run(kill_cmd, ignore_status=True, shell=True)
```

Comment:
Actually, I was intent on putting the nbd interfaces into utils_qemu, because there has already a nbd export class in utils_nbd.py that exports images, but we cannot inherit it or make an extension for too much hard code, then I coded the three interfaces for the backend use. In order to avoid confusing them with the nbdexport class, I put them into utils_qemu

Reply:
Yes, I saw utils_nbd.py that almost is for libvirt layer, but some of them we still could use, maybe we need to add a condition for our qemu layer. 
BTW, from the existing interfaces that focus the qemu utility:
```
get_qemu_version
get_supported_machines_list
get_maxcpus_hard_limit
....
```
Such them,  like getting the qemu capability interfaces.
These don't refer to any qemu instance scenarios, independently.

Reply:
marked this as 'unresolved' for future code change